Fix x-assign unique when index out-of-range of an unpacked struct array

verilator version: >= 5.012 （include the newest version Verilator 5.015 devel rev v5.014-31-g63db60f64） 
(I didn't test the lower version)

bug code:
```verilog
typedef struct {
    logic a;
} Data_t;

module tb (
    input wire clk,
    input wire rst
);
    localparam int SIZE = 20;
    reg[$clog2(SIZE)-1 : 0] ptr;
    Data_t buffer[SIZE];
    reg out;

    always_ff @( posedge clk ) begin
        int i;
        if (rst) begin
            for (i=0;i<SIZE;i=i+1) begin
                buffer[i].a <= 0;
            end
        end
        else begin
            ptr <= (ptr == (SIZE - 1)) ? 0 : (ptr+1);
            out <= buffer[ptr].a;
        end
    end

endmodule
```


with verialtor arguments :
```
--x-assign unique --x-initial unique
```
if SIZE = 32, compile success
but if is 20 or others not a power of 2, compile error
message:
```
Vtb_tb__DepSet_h66f4ffa7__0.cpp:129:33: error: operands to ‘?:’ have different types ‘Vtb_Data_t__struct__0’ and ‘IData’ {aka ‘unsigned int’}
  128 |         vlSelf->__Vdly__out = ((0x13U >= (IData)(vlSelf->__PVT__ptr))
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  129 |                                 ? vlSelf->__PVT__buffer
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~
  130 |                                [vlSelf->__PVT__ptr]
      |                                ~~~~~~~~~~~~~~~~~~~~
  131 |                                 : (IData)(vlSelf->__Vxrand_h9c26aeca__0))
```
generated c++ code is:
```c++
        vlSelf->__Vdly__out = ((0x13U >= (IData)(vlSelf->__PVT__ptr))
                                ? vlSelf->__PVT__buffer
                               [vlSelf->__PVT__ptr]
                                : (IData)(vlSelf->__Vxrand_h9c26aeca__0))
            .__PVT__a;
```
__Vxrand_h9c26aeca__0 is a unsiged instead of struct

if without verilator arguments
then another error message
```
Vtb___024root__DepSet_h00dcc3fb__0.cpp:188:39: error: no match for ‘operator&’ (operand types are ‘bool’ and ‘Vtb_Data_t__struct__0’)
  188 |                                       & vlSelf->tb__DOT__buffer
```

when ptr may go out of the array range
it will generated the
"out = (ptr <= (SZIE-1) ? buffer[ptr] : random_value).a"
c++ code (random_value is a unsigned type)
or
"out = ((ptr <= (SZIE-1) & buffer[ptr]).a" (without verilator arguments)
I think it should be
"out = (ptr <= (SZIE-1) ? buffer[ptr].a : random_value)"

--- a/docs/CONTRIBUTORS
+++ b/docs/CONTRIBUTORS
@@ -189,6 +189,7 @@ Vito Gamberini
 William D. Jones
 Wilson Snyder
 Xi Zhang
+Yan Xu
 Yinan Xu
 Yoda Lee
 Yossi Nivin

--- a/src/V3Unknown.cpp
+++ b/src/V3Unknown.cpp
@@ -471,7 +471,9 @@ class UnknownVisitor final : public VNVisitor {
                 VL_DO_DANGLING(condp->deleteTree(), condp);
             } else if (!lvalue
                        // Making a scalar would break if we're making an array
-                       && !VN_IS(nodep->dtypep()->skipRefp(), NodeArrayDType)) {
+                       && !VN_IS(nodep->dtypep()->skipRefp(), NodeArrayDType)
+                       && !(VN_IS(nodep->dtypep()->skipRefp(), NodeUOrStructDType)
+                       && !VN_CAST(nodep->dtypep()->skipRefp(), NodeUOrStructDType)->packed())) {
                 // ARRAYSEL(...) -> COND(LT(bit<maxbit), ARRAYSEL(...), {width{1'bx}})
                 VNRelinker replaceHandle;
                 nodep->unlinkFrBack(&replaceHandle);

--- a/test_regress/t/t_struct_unpacked_array.pl
+++ b/test_regress/t/t_struct_unpacked_array.pl
@@ -1,0 +1,22 @@
+#!/usr/bin/env perl
+if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# Copyright 2023 by Wilson Snyder. This program is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU
+# Lesser General Public License Version 3 or the Perl Artistic License
+# Version 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+scenarios(simulator => 1);
+
+compile(
+    verilator_flags2=>["--x-assign unique --x-initial unique -Wno-WIDTH -O0"]
+    );
+
+execute(
+    check_finished => 1,
+    );
+
+ok(1);
+1;

--- a/test_regress/t/t_struct_unpacked_array.v
+++ b/test_regress/t/t_struct_unpacked_array.v
@@ -1,0 +1,41 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// This file ONLY is placed under the Creative Commons Public Domain, for
+// any use, without warranty, 2023 by Wilson Snyder.
+// SPDX-License-Identifier: CC0-1.0
+
+
+typedef struct {
+    logic a;
+} Data_t;
+
+module t (/*AUTOARG*/
+    clk
+);
+    input clk;
+    int cyc = 0;
+
+    localparam int SIZE = 20;
+    reg[$clog2(SIZE)-1 : 0] ptr;
+    Data_t buffer[SIZE];
+    Data_t out;
+    reg out1;
+
+    always_ff @( posedge clk ) begin
+        int i;
+        cyc <= cyc + 1;
+        if (cyc == 0) begin
+            for (i=0;i<SIZE;i=i+1) begin
+                buffer[i].a <= 0;
+            end
+        end
+        else begin
+            ptr <= (ptr+1);
+            out <= buffer[ptr];
+            out1 <= buffer[ptr].a;
+            $write("*-* All Finished *-*\n");
+            $finish;
+        end
+    end
+
+endmodule

